rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,267 @@
+- The Teleport proxy translates upgrade hours (below) into a specific time after which all agents should be upgraded.
+- If an agent misses an upgrade window, it will always update immediately.
+
+#### Teleport Resources
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # agent_auto_update allows turning agent updates on or off at the
+  # cluster level. Only turn agent automatic updates off if self-managed
+  # agent updates are in place.
+  agent_auto_update: on|off
+  # agent_update_hour sets the hour in UTC at which clients should update their agents.
+  # The value -1 will set the upgrade time to the current time, resulting in immediate upgrades.
+  agent_update_hour: -1-23
+  # agent_update_jitter_seconds sets a duration in which the upgrade will occur after the hour.
+  # The agent upgrader will pick a random time within this duration in which to upgrade.
+  agent_update_jitter_seconds: 0-MAXINT64
+  
+  [...]
+```
+```
+$ tctl autoupdate update --set-agent-auto-update=off
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-hour=3
+Automatic updates configuration has been updated.
+$ tctl autoupdate update --set-agent-update-jitter-seconds=600
+Automatic updates configuration has been updated.
+```
+
+```yaml
+kind: autoupdate_version
+spec:
+  # agent_version is the version of the agent the cluster will advertise.
+  # Can be auto (match the version of the proxy) or an exact semver formatted
+  # version.
+  agent_version: auto|X.Y.Z
+
+  [...]
+```
+```
+$ tctl autoupdate update --set-agent-version=15.1.1
+Automatic updates configuration has been updated.
+```
+
+Notes:
+- These two resources are separate so that Cloud customers can be restricted from updating `autoupdate_version`, while maintaining control over the rollout.
+
+### Filesystem
+
+```
+$ tree /var/lib/teleport
+/var/lib/teleport
+└── versions
+   ├── 15.0.0
+   │  ├── bin
+   │  │  ├── ...
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     ├── ...
+   │     └── systemd
+   │        └── teleport.service
+   ├── 15.1.1
+   │  ├── bin
+   │  │  ├── ...
+   │  │  ├── teleport-updater
+   │  │  └── teleport
+   │  └── etc
+   │     ├── ...
+   │     └── systemd
+   │        └── teleport.service
+   └── updates.yaml
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport-updater -> /var/lib/teleport/versions/15.0.0/bin/teleport-updater
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+proxy: mytenant.teleport.sh
+enabled: true
+active_version: 15.1.1
+```
+
+### Runtime
+
+The agent-updater will run as a periodically executing systemd service which runs every 10 minutes.
```

Comment:
One option for performing fleet-wide rollbacks would be to avoid advertising earlier, incompatible versions.

If we want to support downgrading fleets that are partial upgraded already, we could advertise another field (`agent_version_downgrade_max`) that is normally set the same as the desired version. If we want to advertise a downgrade, we would set `agent_version_downgrade_max` to the last non-breaking version above the advertised downgrade. For example:
```yaml
agent_version: 14.1.1
agent_version_downgrade_max: 14.2.3
```
Would indicate that the desired version in 14.1.1, but nothing upgraded above 14.2.3 can be downgraded.

If `teleport.yaml` and `/var/lib/teleport` have well-defined schema versions that are available to the agent, it might be better for the upgrader to validate that the version change does not break compatibility with the structures on disk. That would allow downgrades in more scenarios (e.g., teleport never actually ran)

Failure-based rollbacks are already described here: https://github.com/gravitational/teleport/pull/40190/files#diff-4fb277bd3d884c7b045e7efad61111ac4e7957c063e0c0c9281f4e74fd72997fR222

Do you have preferences on which approach we take?

Reply:
Let me know if you prefer the approach above, per-version backups for /var/lib/teleport, or some combination of both.

Holding off on adding either to the RFD until we have consensus.

Reply:
Added per-version backups of `/var/lib/teleport` for now.

Reply:
After chatting this week, it sounds like both `/var/lib/teleport` backups and `agent_version_downgrade_max` could be viable. I would lean towards implementing `agent_version_downgrade_max` first, and adding backups later if needed. This would reduce updater complexity and possible failure states.

Waiting for more feedback before making additional changes.

Reply:
The main risk I see with the backup is if a key rotation starts between the backup and its restoration. I'm not certain if a node that connect to a cluster in the middle of a rotation refreshes its CA keys, but I'm sure that if the rotation is finished, restoring from the backup will not work anymore while keeping the state would have worked.

Reply:
You can descope as much as you want, but if a downgrade makes us restore a backup with credentials that are no longer trusted by the cluster and are no longer trusting the cluster - which will happen if the backup is from before the beginning of a CA rotation, and the rotation was fully completed - the node will break and it'll never heal by itself.

If we have reason to believe that the contents of /var/lib/teleport (which include session uploads, so we should probably also carve those out of backups) will become incompatible across a downgrade, we'd be much better off wiping the state (sans `host_uuid`) and letting the node join again, if we know that the configured token is still valid.

Reply:
Will an `agent_version` lower than `agent_version_downgrade_max` cause a downgrade of older agents down to `agent_version_downgrade_max` exactly, or will those agents simply not downgrade?

Reply:
> Will an agent_version lower than agent_version_downgrade_max cause a downgrade of older agents down to agent_version_downgrade_max exactly, or will those agents simply not downgrade?

I lean towards not downgrading at all. If the downgrade version was acceptable, you'd set `agent_version` to that version instead. The intent of `agent_version_downgrade_max` is to ensure that agents which cannot be downgraded are not broken.

> You can descope as much as you want, but if a downgrade makes us restore a backup with credentials that are no longer trusted by the cluster and are no longer trusting the cluster

I was suggesting descoping backups of `/var/lib/teleport` entirely, due to the added complexity. It's not a capability that we have today, so I don't see a downside to shipping the rest of the RFD without it. `agent_version_downgrade_max` will provide some protection until we have a more comprehensive solution. Maybe it would be simpler to stop changing `/var/lib/teleport` in backwards-incompatible ways? Or provide that guarantee for a critical subset of the structure?

Reply:
We also know if a CA rotation is in-progress on the control plane side. We can just not perform an upgrade during an active CA rotation. A user could kick off a CA rotation at the exact same time we perform a CA rotation but the risk feels minimal. In the future we could even lock out allowing CA rotation while we perform the upgrade?

I do feel strongly that we do need the ability to perform rollbacks in the case of catastrophic failure of a failed agent upgrade.

Reply:
+1 for just not breaking the state and allowing rollbacks. There's some risk of downgrade attack but I'm not sure what would be the best mitigation.

The biggest risk with the "no breaking change and no backup" approach is if a new teleport version mangles the state or has non forward-compatible changes.

But I feel it should be teleport's responsibility to support loading old state, or ignoring new unknown fields on a resource in its state instead of crashing.